src/main/host/syscall/socket.rs

```diff
@@ -0,0 +1,578 @@
+
+    // clean up in case of error
+    match write_res {
+        Ok(_) => Ok(0.into()),
+        Err(e) => {
+            EventQueue::queue_and_run(|event_queue| {
+                // ignore any errors when closing
+                ctx.process
+                    .deregister_descriptor(fd_1)
+                    .unwrap()
+                    .close(ctx.host.chost(), event_queue);
+                ctx.process
+                    .deregister_descriptor(fd_2)
+                    .unwrap()
+                    .close(ctx.host.chost(), event_queue);
+            });
+            Err(e.into())
+        }
+    }
+}
+
+/// Returns a nix socket address object where only the family is set.
+fn empty_sockaddr(family: nix::sys::socket::AddressFamily) -> nix::sys::socket::SockAddr {
+    let family = family as libc::sa_family_t;
+    let mut addr: nix::sys::socket::sockaddr_storage = unsafe { std::mem::zeroed() };
+    addr.ss_family = family;
+    // the size of ss_family will be 2 bytes on linux
+    nix::sys::socket::sockaddr_storage_to_addr(&addr, 2).unwrap()
+}
+
+/// Copy the socket address to the plugin. Will return an error if either the address or address
+/// length pointers are NULL. The plugin's address length will be updated to store the size of the
+/// socket address, even if greater than the provided buffer size. If the address is `None`, the
+/// plugin's address length will be set to 0.
+fn write_sockaddr(
+    process: &mut Process,
+    addr: Option<nix::sys::socket::SockAddr>,
+    plugin_addr: PluginPtr,
+    plugin_addr_len: TypedPluginPtr<libc::socklen_t>,
+) -> Result<(), SyscallError> {
+    if plugin_addr.is_null() || plugin_addr_len.is_null() {
+        return Err(Errno::EFAULT.into());
+    }
+
+    let addr = match addr {
+        Some(x) => x,
+        None => {
+            process.memory_mut().copy_to_ptr(plugin_addr_len, &[0])?;
+            return Ok(());
+        }
+    };
+
+    let (from_addr, from_len) = addr.as_ffi_pair();
+
+    let plugin_addr_len = {
+        let mut plugin_addr_len = process.memory_mut().memory_ref_mut(plugin_addr_len)?;
+        let plugin_addr_len_value = plugin_addr_len.get_mut(0).unwrap();
+
+        // keep a copy before we change it
+        let plugin_addr_len_copy = *plugin_addr_len_value;
+
+        *plugin_addr_len_value = from_len;
+
+        plugin_addr_len.flush()?;
+        plugin_addr_len_copy
+    };
+
+    // return early if the address length is 0
+    if plugin_addr_len == 0 {
+        return Ok(());
+    }
+
+    let plugin_addr = TypedPluginPtr::new::<u8>(plugin_addr, (plugin_addr_len).try_into().unwrap());
+    let mut plugin_addr = process.memory_mut().memory_ref_mut_uninit(plugin_addr)?;
+
+    // make sure that we have a real libc sockaddr before converting to a u8 pointer
+    let from_addr: &libc::sockaddr = from_addr;
+
+    let len_to_copy = std::cmp::min(from_len.try_into().unwrap(), plugin_addr.len());
+    unsafe {
```

Comment:
I think `process.memory_mut().copy_to_ptr` would be a bit simpler here. It also potentially saves an allocation and copy through an intermediate buffer in the case the address isn't mapped into shadow.

Reply:
Updated.